src/backend/catalog/test/storage_pending_deletes_test.c

```diff
@@ -0,0 +1,525 @@
+/*-------------------------------------------------------------------------
+ *
+ * storage_pending_deletes_test.c
+ *	  code to test functionality from storage_pending_deletes.c
+ *
+ * Copyright (c) 2025 Greengage Community
+ *
+ *	  src/backend/catalog/test/storage_pending_deletes_test.c
+ *
+ *-------------------------------------------------------------------------
+ */
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include "cmockery.h"
+
+#include "catalog/storage_pending_deletes.h"
+#include "storage/pg_shmem.h"
+#include "storage/proc.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
+
+enum
+{
+	TEST_TABLESPACE_OID1 = 11111,
+	TEST_TABLESPACE_OID2 = 11112,
+
+	TEST_DB_OID1 = 11121,
+	TEST_DB_OID2 = 11122,
+
+	TEST_REL_OID1 = 11211,
+	TEST_REL_OID2 = 11212,
+
+	TEST_XID1 = 10,
+	TEST_XID2 = TEST_XID1 + 1,
+	TEST_XID3 = TEST_XID1 + 3,
+	TEST_XID4 = TEST_XID1 + 8,
+};
+
+/* Don't try to read a non-existent postmaster.pid file */
+void		__wrap_AddToDataDirLockFile(int target_line, const char *str);
+void
+__wrap_AddToDataDirLockFile(int target_line, const char *str)
+{
+}
+
+/* Dump without additions */
+static void
+test_empty(void **state)
+{
+	assert_true(PdlXLogShmemDump() == NULL);
+}
+
+/* Add single pending delete node */
+static void
+test_1(void **state)
+{
+	const RelFileNodePendingDelete relnode =
+	{
+		.node =
+		{
+			.spcNode = TEST_TABLESPACE_OID1,
+			.dbNode  = TEST_DB_OID1,
+			.relNode = TEST_REL_OID1
+		},
+		.relstorage = RELSTORAGE_HEAP
+	};
+
+	dsa_pointer p = PdlShmemAdd(&relnode, TEST_XID1);
+	PendingRelXactDeleteArray *arr = PdlXLogShmemDump();
+
+	assert_true(arr != NULL);
+	assert_int_equal(arr->count, 1);
+	assert_int_equal(arr->array[0].xid, TEST_XID1);
+	assert_memory_equal(&arr->array[0].relnode, &relnode, sizeof(relnode));
+
+	/* clean up */
+	pfree(arr);
+	PdlShmemRemove(p);
+}
+
+/* Add nodes and remove the first one before dump */
+static void
+test_remove_fisrt(void **state)
+{
+	RelFileNodePendingDelete relnode =
+	{
+		.node =
+		{
+			.spcNode = TEST_TABLESPACE_OID1,
+			.dbNode = TEST_DB_OID1,
+			.relNode = TEST_REL_OID1
+		}
+	};
+
+	dsa_pointer p[5];
+
+	p[0] = PdlShmemAdd(&relnode, TEST_XID1);
+	relnode.node.spcNode = TEST_TABLESPACE_OID2;
+	p[1] = PdlShmemAdd(&relnode, TEST_XID2);
+	relnode.node.dbNode = TEST_DB_OID2;
+	p[2] = PdlShmemAdd(&relnode, TEST_XID3);
+	relnode.node.relNode = TEST_REL_OID2;
+	p[3] = PdlShmemAdd(&relnode, TEST_XID1);
+	relnode.node.spcNode = TEST_TABLESPACE_OID1;
+	p[4] = PdlShmemAdd(&relnode, TEST_XID1);
```

Comment:
Let's insert empty line between inserts to improve readability (here and in other tests)
```suggestion
	p[0] = PdlShmemAdd(&relnode, TEST_XID1);
	
	relnode.node.spcNode = TEST_TABLESPACE_OID2;
	p[1] = PdlShmemAdd(&relnode, TEST_XID2);
	
	relnode.node.dbNode = TEST_DB_OID2;
	p[2] = PdlShmemAdd(&relnode, TEST_XID3);
	
	relnode.node.relNode = TEST_REL_OID2;
	p[3] = PdlShmemAdd(&relnode, TEST_XID1);
	
	relnode.node.spcNode = TEST_TABLESPACE_OID1;
	p[4] = PdlShmemAdd(&relnode, TEST_XID1);
```

The same between checking of `arr->array[0]` and `arr->array[1]`, etc...

Reply:
Done